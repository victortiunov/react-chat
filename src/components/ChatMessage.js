import React from 'react';
import classnames from 'classnames';
import moment from 'moment';
import { withStyles } from 'material-ui';
import Typography from 'material-ui/Typography';
import Paper from 'material-ui/Paper';
import Avatar from './Avatar';

import userName from '../utils/user-name';
import getColor from '../utils/color-from';

const styles = theme => ({
	messageWrapper: {
		display: 'flex',
		justifyContent: 'flex-start',
		alignItems: 'center',
		padding: `${theme.spacing.unit}px ${theme.spacing.unit * 3}px`,
	},
	messageWrappperFromMe: {
		justifyContent: 'flex-end',
	},
	message: {
		maxWidth: '70%',
		minWidth: '10%',
		padding: theme.spacing.unit,
		marginLeft: theme.spacing.unit * 2,
	},
	messageFromMe: {
		marginRight: theme.spacing.unit * 2,
		backgroundColor: '#e6dcff',
	},
	statusMessage: {
		justifyContent: 'center',
	},
	statusMessageUser: {
		display: 'inline',
	},
});

const ChatMessage = ({
	classes, user, sender, content, createdAt, statusMessage,
}) => {
	// eslint-disable-next-line
	const isMessageFromMe = sender._id === user._id;
	const senderName = userName(sender.username, sender.firstName, sender.lastName);

	if (statusMessage) {
		return (
			<div className={classnames(classes.messageWrapper, classes.statusMessage)}>
				<Typography
					className={classes.statusMessageUser}
					style={{ color: getColor(senderName) }}
				>
					{senderName}
				</Typography>
				&nbsp;
				<Typography variant="caption">
					{content} {moment(createdAt).fromNow()}
				</Typography>
			</div>
		);
	}

	const userAvatar = (
		<Avatar colorFrom={senderName}>
			{senderName}
		</Avatar>
	);

	return (
		<div className={classnames(
			classes.messageWrapper,
			isMessageFromMe && classes.messageWrappperFromMe,
		)}
		>
			{!isMessageFromMe && userAvatar}
			<Paper className={classnames(
				classes.message,
				isMessageFromMe && classes.messageFromMe,
			)}
			>
				<Typography variant="caption">
					{senderName}
				</Typography>
				<Typography variant="body1">
					{content}
					<Typography variant="caption" component="span">
						{moment(createdAt).fromNow()}
					</Typography>
				</Typography>
			</Paper>
			{isMessageFromMe && userAvatar}
		</div>
	);
};

export default withStyles(styles)(ChatMessage);
