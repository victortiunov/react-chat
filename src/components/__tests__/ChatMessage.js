/* eslint-env jest */
import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import ChatMessage from '../ChatMessage';

// Return a fixed timestamp when moment().fromNow() is called
jest.mock('moment', () => () => ({ fromNow: () => '37 days ago' }));

const mockProps = {
  user: {
    _id: '373737',
  },
  sender: {
    _id: '373737',
    username: 'hpotter',
    firstName: 'Harry',
    lastName: 'Potter',
  },
  content: 'Expecto Patronum!',
  createdAt: '1994-06-16T21:53:23.200Z',
  statusMessage: false,
};

describe('<ChatMessage />', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<ChatMessage {...mockProps} />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders message from current user correctly', () => {
    const tree = renderer.create(<ChatMessage {...mockProps} />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders message from other user correctly', () => {
    const content = 'Dobby is a free elf!';
    const tree = renderer
      .create(<ChatMessage
        {...mockProps}
        sender={{
            _id: '737373',
            username: 'dobby',
            firstName: 'Dobby',
            lastName: 'Elf',
          }}
        content={content}
      />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders status message correctly', () => {
    const content = 'joined';
    const tree = renderer
      .create(<ChatMessage {...mockProps} content={content} statusMessage />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
