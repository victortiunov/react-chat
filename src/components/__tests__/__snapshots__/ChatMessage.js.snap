// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChatMessage /> renders message from current user correctly 1`] = `
<div
  className="ChatMessage-messageWrapper-61 ChatMessage-messageWrappperFromMe-62"
>
  <div
    className="MuiPaper-root-67 MuiPaper-elevation2-71 MuiPaper-rounded-68 ChatMessage-message-63 ChatMessage-messageFromMe-64"
  >
    <span
      className="MuiTypography-root-94 MuiTypography-caption-104"
    >
      Harry Potter
    </span>
    <p
      className="MuiTypography-root-94 MuiTypography-body1-103"
    >
      Expecto Patronum!
      <span
        className="MuiTypography-root-94 MuiTypography-caption-104"
      >
        37 days ago
      </span>
    </p>
  </div>
  <div
    className="MuiAvatar-root-118 MuiAvatar-colorDefault-119"
    style={
      Object {
        "backgroundColor": "#009688",
      }
    }
  >
    HP
  </div>
</div>
`;

exports[`<ChatMessage /> renders message from other user correctly 1`] = `
<div
  className="ChatMessage-messageWrapper-61"
>
  <div
    className="MuiAvatar-root-118 MuiAvatar-colorDefault-119"
    style={
      Object {
        "backgroundColor": "#e91e63",
      }
    }
  >
    DE
  </div>
  <div
    className="MuiPaper-root-67 MuiPaper-elevation2-71 MuiPaper-rounded-68 ChatMessage-message-63"
  >
    <span
      className="MuiTypography-root-94 MuiTypography-caption-104"
    >
      Dobby Elf
    </span>
    <p
      className="MuiTypography-root-94 MuiTypography-body1-103"
    >
      Dobby is a free elf!
      <span
        className="MuiTypography-root-94 MuiTypography-caption-104"
      >
        37 days ago
      </span>
    </p>
  </div>
</div>
`;

exports[`<ChatMessage /> renders status message correctly 1`] = `
<div
  className="ChatMessage-messageWrapper-61 ChatMessage-statusMessage-65"
>
  <p
    className="MuiTypography-root-94 MuiTypography-body1-103 ChatMessage-statusMessageUser-66"
    style={
      Object {
        "color": "#009688",
      }
    }
  >
    Harry Potter
  </p>
  Â 
  <span
    className="MuiTypography-root-94 MuiTypography-caption-104"
  >
    joined
     
    37 days ago
  </span>
</div>
`;
